### PR TITLE
[SPARK-41221][SQL] Add the error class `INVALID_FORMAT`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to move out the error sub-classes related to number format issues from `DATATYPE_MISMATCH` to their own error class `INVALID_FORMAT`. Also, the PR deletes the error class `INVALID_LIKE_PATTERN`, and moves their sub-classes to `INVALID_FORMAT`.

### Why are the changes needed?
To improve usability of Spark SQL and to don't confuse users by error class names related to values but to to its types.

### Does this PR introduce _any_ user-facing change?
Yes, the PR changes some user-facing errors.

### How was this patch tested?
By running the affected test suites:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
$ build/sbt "core/testOnly *SparkThrowableSuite"
$ build/sbt "test:testOnly *StringExpressionsSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srielau @srielau @itholic @panbingkun @LuciferYang Could you have a look at this PR, please.

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -687,6 +647,63 @@
     ],
     "sqlState" : "42000"
   },
+  "INVALID_FORMAT" : {
```
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:650]: Do we want this to be generic or specific to numbers.
I'm specifically wondering about DATE FORMAT, INTERVAL FORMAT, .. What have we/will we do with these?

**Review by srielau — APPROVED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (srielau) [reply core/src/main/resources/error/error-classes.json:650]: Basically: Should we use INVALID_NUMBER_FORMAT, INVALID_DATETIME_FROMAT, .... or just this and bring them all in?

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:650]: > Do we want this to be generic or specific to numbers.

I want to believe that we can make it as a generic error class. I have already merged `LIKE ...` errors into it.

I think there is the common part where we declare that the format is invalid and show it.

**Review by cloud-fan — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srielau @cloud-fan @LuciferYang for review.

**Pull request closed without merging**